File: apps/flutter-prototype/test/recovered_conversation_summary_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:meshx_flutter_probe/core/models.dart';
import 'package:meshx_flutter_probe/core/recovered_conversation_summary.dart';

ChatMessage message(
  int sequence, {
  int sender = 8,
  String content = 'verified',
  String type = 'text',
  bool burn = false,
}) => ChatMessage(
  messageId: 'm$sequence',
  clientMsgId: '',
  conversationId: 'group:21',
  fromUserId: sender,
  content: content,
  contentType: type,
  sequence: sequence,
  createdAt: DateTime(2026),
  isBurn: burn,
);
void main() {
  test(
    'verified summary counts peers after own read position and orders by sequence',
    () {
      final summary = recoveredConversationSummary(
        [message(3, sender: 7, content: 'own'), message(1), message(2)],
        [],
        7,
        1,
      );
      expect(summary, (preview: 'own', unread: 1, timestamp: DateTime(2026)));
    },
  );
  test('terminal latest summary never exposes stale body', () {
    for (final entry in {
      'RECALLED': '这条消息已撤回',
      'BURNED': '这条消息已焚毁',
      'UNAVAILABLE': '这条消息已不可用',
    }.entries) {
      final summary = recoveredConversationSummary(
        [message(1)],
        [
          {'state': entry.key, 'messageSequence': '2'},
        ],
        7,
        0,
      );
      expect(summary, (preview: entry.value, unread: 1, timestamp: null));
    }
  });
  test(
    'burn and unknown media have safe labels; missing baseline does not invent unread',
    () {
      expect(
        recoveredConversationSummary([message(1, burn: true)], [], 7, null),
        (preview: '阅后即焚消息', unread: 0, timestamp: DateTime(2026)),
      );
      expect(
        recoveredConversationSummary(
          [message(1, type: 'image')],
          [],
          7,
          null,
        ).preview,
        '图片',
      );
      expect(
        recoveredConversationSummary(
          [message(1, type: 'unknown')],
          [],
          7,
          null,
        ).preview,
        '暂不预览此类消息',
      );
      expect(recoveredConversationSummary([], [], 7, 0), (
        preview: '',
        unread: 0,
        timestamp: null,
      ));
    },
  );
}

File: apps/web/tests/recovered-conversation-summary.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'
import {createTsLoader,plain} from './helpers/load-ts.mjs'
const {recoveredConversationSummaries:summarize}=createTsLoader()('packages/domain-ts/src/recoveredConversationSummary.ts')
const row={conversationId:'group:21',messageId:'peer',sequence:1,state:'NORMAL',fromUserId:8,content:'verified',contentType:'text',timestamp:'2026-09-20T00:00:00'}
test('verified summaries count peer history after own read position and use sequence ordering',()=>{
 const result=summarize([{...row,sequence:3,messageId:'own',fromUserId:7,content:'own'},row,{...row,messageId:'peer2',sequence:2}],7,new Map([['group:21',1]])).get('group:21')
 assert.deepEqual(plain(result),{preview:'own',unreadCount:1,lastSequence:3,timestamp:row.timestamp})
})
test('terminal latest summary has no stale body or fabricated timestamp',()=>{
 for(const [state,label] of [['RECALLED','这条消息已撤回'],['BURNED','这条消息已焚毁'],['UNAVAILABLE','这条消息已不可用']]) {
  const result=summarize([row,{conversationId:row.conversationId,messageId:'terminal',sequence:2,state}],7,new Map([['group:21',0]])).get('group:21')
  assert.deepEqual(plain(result),{preview:label,unreadCount:1,lastSequence:2})
 }
})
test('burn bodies and unsupported media never become list previews; missing baseline does not invent unread',()=>{
 for(const [patch,label] of [[{isBurn:true},'阅后即焚消息'],[{contentType:'image'},'图片'],[{contentType:'unknown'},'暂不预览此类消息']]) {
  const result=summarize([{...row,...patch}],7,new Map()).get('group:21')
  assert.equal(result.preview,label);assert.equal(result.unreadCount,0)
 }
 assert.equal(summarize([],7,new Map()).size,0)
})
